SUP-4735, SUP-4736, SUP-4737, SUP-4738: Add ArgoCD Deployment Plugin Implementation

Mykematt opened the pull request:
Complete implementation of ArgoCD deployment plugin with comprehensive features

Commit:
Add complete ArgoCD deployment plugin implementation

- Deploy and rollback ArgoCD applications
- Auto/manual rollback modes with health monitoring
- Cross-platform compatibility (Mac, Linux, K8s, Elastic CI)
- Log collection, artifact upload, and Slack notifications
- Comprehensive documentation and test suite
- Modular architecture following Buildkite standards

tomowatt commented on lib/modules/notifications.bash:
```diff
@@ -0,0 +1,251 @@
+      - slack:
+          channels:
+            - "$slack_channel"
+          message: "$escaped_message"
```
With Clusters and Queues being a thing, uploading Steps with/without `agents` attributes may not work as there might not be "default" queue to fall back on that could run this Step.

tomowatt commented on lib/modules/rollback.bash:
```diff
@@ -0,0 +1,395 @@
+          - label: "❌ Accept Failure (No Rollback)"
+            value: "accept"
+
+  - label: "Execute User Decision"
+    depends_on: "deployment-failed-choose-action"
+    agents:
+      queue: kubernetes
+    command: |
```
In the similar vein to my above comment, we cannot expect users to have a specific Queue to run a Step, so might need a bit of thought to how to approach uploading Steps via the Plugin so they can run.

Commit:
Fix queue targeting for uploaded steps

- Add dynamic queue resolution using BUILDKITE_AGENT_META_DATA_QUEUE
- Ensures manual rollback and notification steps run on same queue as deployment
- Addresses review feedback about agent targeting in uploaded pipeline steps

Changes:
- rollback.bash: Add queue targeting to 'Execute User Decision' step
- notifications.bash: Add queue targeting to Slack notification step
- Both use actual agent queue instead of hardcoded values

Mykematt commented:
Thanks @tomowatt 

I've pushed a commit that fixed this. The fix involves using local vars to fetch current running agent queue. Then the local var is used in the uploaded pipelines. Here are tests that confirm that both execute_decision and slack_notification steps are now handled by whatever agent users used to run the build: 
- Default queue (Elastic CI stack agent): https://buildkite.com/olabuildkite/argocd-plugin-test/builds/261/steps/table
- Kubernetes queue (Agent-stack-k8s): https://buildkite.com/olabuildkite/argocd-plugin-test/builds/262/steps/table
- mac queue (my local mac agent): https://buildkite.com/olabuildkite/argocd-plugin-test/builds/263/steps/table

jeremybumsted commented on lib/modules/notifications.bash:
```diff
@@ -38,6 +38,9 @@ send_notification() {
     local to_revision="$4"
     
+    # Get current queue for proper agent targeting
+    local current_queue="$BUILDKITE_AGENT_META_DATA_QUEUE"
```
@Mykematt a question while reviewing this: Is the intention here to ensure that the steps are assigned to the same queue that the uploading job is running from? Does it make sense to allow the queue to be configurable by plugin config (and fall back to this as a sensible default?)

Otherwise LGTM :)
